staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go

```diff
@@ -149,290 +169,196 @@ func PatchResource(r rest.Patcher, scope RequestScope, admit admission.Interface
+	switch p.patchType {
 	case types.JSONPatchType:
-		patchObj, err := jsonpatch.DecodePatch(patchJS)
+		patchObj, err := jsonpatch.DecodePatch(p.patchJS)
 		if err != nil {
-			return nil, nil, err
-		}
-		if patchedObjJS, err = patchObj.Apply(originalObjJS); err != nil {
-			return nil, nil, err
+			return nil, err
 		}
+		return patchObj.Apply(versionedJS)
 	case types.MergePatchType:
-		if patchedObjJS, err = jsonpatch.MergePatch(originalObjJS, patchJS); err != nil {
-			return nil, nil, err
-		}
-	case types.StrategicMergePatchType:
-		if patchedObjJS, err = strategicpatch.StrategicMergePatch(originalObjJS, patchJS, versionedObj); err != nil {
-			return nil, nil, err
-		}
+		return jsonpatch.MergePatch(versionedJS, p.patchJS)
 	default:
 		// only here as a safety net - go-restful filters content-type
-		return nil, nil, fmt.Errorf("unknown Content-Type header for patch: %v", patchType)
+		return nil, fmt.Errorf("unknown Content-Type header for patch: %v", p.patchType)
 	}
-	if err := runtime.DecodeInto(codec, patchedObjJS, objToUpdate); err != nil {
-		return nil, nil, err
+}
+
+func (p *jsonPatcher) subsequentPatchAttempt(currentObject runtime.Object, currentResourceVersion string) (runtime.Object, error) {
+	return subsequentPatchLogic(p.patcher, p, currentObject, currentResourceVersion)
+}
+
+// Return a fresh strategic patch map if needed for conflict retries. We have
+// to rebuild it each time we need it, because the map gets mutated when being
+// applied.
+func (p *jsonPatcher) originalStrategicMergePatch() (map[string]interface{}, error) {
+	if p.originalPatchBytes == nil {
+		// Compute once. Compute here instead of in the first patch
+		// attempt because this isn't needed unless there's actually a
+		// conflict.
+		var err error
+		p.originalPatchBytes, err = strategicpatch.CreateTwoWayMergePatch(p.originalObjJS, p.originalPatchedObjJS, p.schemaReferenceObj)
+		if err != nil {
+			return nil, interpretPatchError(err)
+		}
+	}
+
+	// Return a fresh map every time
+	originalPatchMap := make(map[string]interface{})
+	if err := json.Unmarshal(p.originalPatchBytes, &originalPatchMap); err != nil {
+		return nil, errors.NewBadRequest(err.Error())
 	}
-	return
+	return originalPatchMap, nil
+}
+
+func (p *jsonPatcher) computeStrategicMergePatch(currentObject runtime.Object, _ map[string]interface{}) (map[string]interface{}, error) {
+	// Compute current patch.
+	currentObjJS, err := runtime.Encode(p.codec, currentObject)
+	if err != nil {
+		return nil, err
+	}
+	currentPatch, err := strategicpatch.CreateTwoWayMergePatch(p.originalObjJS, currentObjJS, p.schemaReferenceObj)
+	if err != nil {
+		return nil, interpretPatchError(err)
+	}
+	currentPatchMap := make(map[string]interface{})
+	if err := json.Unmarshal(currentPatch, &currentPatchMap); err != nil {
+		return nil, errors.NewBadRequest(err.Error())
+	}
+
+	return currentPatchMap, nil
+}
+
+type smpPatcher struct {
+	*patcher
```

Comment:
inlining the patcher in the mechanism makes it hard to tell which fields of smpPatcher are safe to mutate during the operation and which are supposed to be immutable